File: packages/SyncCore/Sources/MacReceiverKit/SyncServerSession.swift
import Foundation
import SyncCore

public enum SyncServerSessionError: Error, Equatable, LocalizedError, Sendable {
    case idleTimedOut
    case integrityFailureLimitExceeded
    case invalidChunk
    case openingTimedOut
    case protocolViolation
    case sessionRejected(String)

    public var errorDescription: String? {
        switch self {
        case .idleTimedOut:
            "The sender stopped sending before the session finished; "
                + "the iPhone was most likely locked or suspended."
        case .integrityFailureLimitExceeded:
            "The same resource failed integrity verification twice."
        case .invalidChunk:
            "The sender provided an invalid or out-of-order chunk."
        case .openingTimedOut:
            "The incoming connection did not open a valid session before the deadline."
        case .protocolViolation:
            "The sender violated the sync protocol."
        case let .sessionRejected(message):
            message
        }
    }
}

public actor SyncServerSession {
    public typealias AcceptedHandler = @Sendable () async -> Void
    public typealias EventHandler = @Sendable (OperationLogEvent) async -> Void

    public static let defaultOpeningTimeout: Duration = .seconds(15)
    // An open session that stops producing frames is almost always an iPhone
    // that got locked or suspended mid-transfer. Without this bound the
    // receiver would block on `receive()` until TCP keepalive gives up
    // (~2 hours on Darwin) and reject every following connection.
    public static let defaultIdleTimeout: Duration = .seconds(45)

    private let manifest: ManifestStore
    private let writer: DestinationWriter
    private var integrityFailures: [String: Int] = [:]
    private var idleTimeout: Duration = SyncServerSession.defaultIdleTimeout

    public init(manifest: ManifestStore, writer: DestinationWriter) {
        self.manifest = manifest
        self.writer = writer
    }

    public func run(
        connection: FramedConnection,
        openingTimeout: Duration = SyncServerSession.defaultOpeningTimeout,
        idleTimeout: Duration = SyncServerSession.defaultIdleTimeout,
        onAccepted: AcceptedHandler? = nil,
        onEvent: EventHandler? = nil
    ) async throws -> SyncSummary {
        defer { connection.cancel() }
        self.idleTimeout = idleTimeout

        try await openSession(
            connection: connection,
            timeout: openingTimeout,
            onEvent: onEvent
        )
        await onAccepted?()

        var summary = SyncSummary.zero
        while true {
            let frame = try await receive(from: connection)
            let message = try frame.controlMessage()
            switch message {
            case .session(.finished):
                try await connection.send(try SyncFrame.control(
                    .result(.sessionCompleted(summary)),
                    requestID: frame.requestID
                ))
                await onEvent?(OperationLogEvent(
                    level: .success,
                    category: "Session",
                    message: "Completed: \(summary.added) added, "
                        + "\(summary.existing) already present, "
                        + "\(summary.notLocal) not local, \(summary.failed) failed."
                ))
                return summary
            case let .offer(offer):
                try await handle(
                    offer: offer,
                    requestID: frame.requestID,
                    connection: connection,
                    summary: &summary,
                    onEvent: onEvent
                )
            default:
                throw SyncServerSessionError.protocolViolation
            }
        }
    }

    private func openSession(
        connection: FramedConnection,
        timeout: Duration,
        onEvent: EventHandler?
    ) async throws {
        let deadline = SessionOpeningDeadline()
        let timeoutTask = Task {
            do {
                try await Task.sleep(for: timeout)
            } catch {
                return
            }
            guard await deadline.markTimedOut() else { return }
            connection.cancel()
        }
        defer { timeoutTask.cancel() }

        do {
            try await connection.start()
            try Task.checkCancellation()

            try await acceptOpeningRequest(
                connection: connection,
                onEvent: onEvent
            )
            guard await deadline.markAccepted() else {
                throw SyncServerSessionError.openingTimedOut
            }
        } catch {
            if await deadline.didTimeOut {
                throw SyncServerSessionError.openingTimedOut
            }
            throw error
        }
    }

    private func acceptOpeningRequest(
        connection: FramedConnection,
        onEvent: EventHandler?
    ) async throws {
        let openingFrame = try await connection.receive()
        guard openingFrame.kind == .session,
              case let .session(.request(albumID, albumName, requestedBinding)) = try openingFrame.controlMessage()
        else {
            throw SyncServerSessionError.protocolViolation
        }
        await onEvent?(OperationLogEvent(
            level: .info,
            category: "Session",
            message: "Opening album “\(albumName)”."
        ))
        let acceptedAlbum: AcceptedAlbum
        do {
            acceptedAlbum = try await manifest.acceptSession(
                albumID: albumID,
                albumName: albumName,
                requestedBindingID: requestedBinding
            )
        } catch ManifestStoreError.sourceBindingMismatch {
            await onEvent?(OperationLogEvent(
                level: .error,
                category: "Session",
                message: "Rejected album “\(albumName)”: source binding mismatch."
            ))
            try await connection.send(try SyncFrame.control(
                .session(.rejected(reason: "source-binding-mismatch")),
                requestID: openingFrame.requestID
            ))
            throw SyncServerSessionError.sessionRejected(
                "The iPhone source binding does not match this destination."
            )
        }
        do {
            try await writer.prepareAlbumDirectory(
                named: acceptedAlbum.destinationFolderName
            )
        } catch {
            await onEvent?(OperationLogEvent(
                level: .error,
                category: "Destination",
                message: "Could not prepare album “\(albumName)”: "
                    + error.localizedDescription
            ))
            try await connection.send(try SyncFrame.control(
                .session(.rejected(reason: "destination-unavailable")),
                requestID: openingFrame.requestID
            ))
            throw SyncServerSessionError.sessionRejected(
                "The album destination folder is unavailable: \(error.localizedDescription)"
            )
        }
        try await connection.send(try SyncFrame.control(
            .session(.accepted(sourceBindingID: acceptedAlbum.sourceBindingID)),
            requestID: openingFrame.requestID
        ))
        await onEvent?(OperationLogEvent(
            level: .success,
            category: "Session",
            message: "Accepted album “\(albumName)” in folder "
                + "“\(acceptedAlbum.destinationFolderName)”."
        ))
    }

    private func handle(
        offer: ResourceOffer,
        requestID: UUID,
        connection: FramedConnection,
        summary: inout SyncSummary,
        onEvent: EventHandler?
    ) async throws {
        let resourceName = offer.descriptor.originalFilename
        await onEvent?(OperationLogEvent(
            level: .info,
            category: "Resource",
            message: "Offered “\(resourceName)” "
                + "(\(offer.descriptor.expectedSize) bytes)."
        ))
        do {
            switch try await writer.begin(offer) {
            case .adopted:
                integrityFailures.removeValue(forKey: offer.resourceID)
                summary.existing += 1
                try await connection.send(try SyncFrame.control(
                    .decision(.skip),
                    requestID: requestID
                ))
                await onEvent?(OperationLogEvent(
                    level: .info,
                    category: "Resource",
                    message: "Skipped “\(resourceName)”; already present."
                ))
                return
            case let .transfer(offset, _):
                let decision: TransferDecision = offset == 0
                    ? .start(offset: 0)
                    : .resume(offset: offset)
                try await connection.send(try SyncFrame.control(
                    .decision(decision),
                    requestID: requestID
                ))
                await onEvent?(OperationLogEvent(
                    level: .info,
                    category: "Resource",
                    message: offset == 0
                        ? "Receiving “\(resourceName)”."
                        : "Resuming “\(resourceName)” at byte \(offset)."
                ))
                try await receiveBytes(
                    offer: offer,
                    requestID: requestID,
                    startingOffset: offset,
                    connection: connection
                )
            }

            let committedURL = try await writer.commit(
                expectedHash: offer.descriptor.contentHash
            )
            let relativePath = String(
                committedURL.path.dropFirst(
                    committedURL.deletingLastPathComponent().path.count + 1
                )
            )
            let snapshot = try await manifest.snapshot(resourceID: offer.resourceID)
            let committedRelativePath = snapshot?.finalRelativePath ?? relativePath
            integrityFailures.removeValue(forKey: offer.resourceID)
            summary.added += 1
            try await connection.send(try SyncFrame.control(
                .result(.committed(relativePath: committedRelativePath)),
                requestID: requestID
            ))
            await onEvent?(OperationLogEvent(
                level: .success,
                category: "Resource",
                message: "Committed “\(resourceName)” to "
                    + "“\(committedRelativePath)”."
            ))
        } catch {
            await writer.abort()
            let isIntegrityMismatch = (error as? DestinationWriterError) == .integrityMismatch
            let retryableIntegrityFailure: Bool
            if isIntegrityMismatch {
                let failureCount = (integrityFailures[offer.resourceID] ?? 0) + 1
                integrityFailures[offer.resourceID] = failureCount
                retryableIntegrityFailure = failureCount == 1
            } else {
                retryableIntegrityFailure = false
            }
            if !isIntegrityMismatch || !retryableIntegrityFailure {
                summary.failed += 1
            }
            let failure = failureResult(
                for: error,
                retryableIntegrityFailure: retryableIntegrityFailure
            )
            await onEvent?(OperationLogEvent(
                level: retryableIntegrityFailure ? .warning : .error,
                category: "Resource",
                message: "Failed “\(resourceName)”: \(error.localizedDescription)"
                    + (retryableIntegrityFailure ? " Retrying once." : "")
            ))
            try? await connection.send(try SyncFrame.control(
                .result(failure),
                requestID: requestID
            ))
            if isIntegrityMismatch && !retryableIntegrityFailure {
                throw SyncServerSessionError.integrityFailureLimitExceeded
            }
            if error is SyncServerSessionError {
                throw error
            }
            if !isIntegrityMismatch {
                throw SyncServerSessionError.sessionRejected(
                    "Resource transfer failed: \(error.localizedDescription)"
                )
            }
        }
    }

    /// Receives the next frame, bounding the wait so a sender that vanishes
    /// (locked iPhone, suspended app, LAN drop) ends the session instead of
    /// holding the receiver's single active connection slot open forever.
    private func receive(from connection: FramedConnection) async throws -> SyncFrame {
        let idleTimeout = idleTimeout
        return try await withThrowingTaskGroup(of: SyncFrame.self) { group in
            group.addTask {
                try await connection.receive()
            }
            group.addTask {
                try await Task.sleep(for: idleTimeout)
                connection.cancel()
                throw SyncServerSessionError.idleTimedOut
            }
            defer { group.cancelAll() }
            guard let frame = try await group.next() else {
                throw SyncServerSessionError.idleTimedOut
            }
            return frame
        }
    }

    private func receiveBytes(
        offer: ResourceOffer,
        requestID: UUID,
        startingOffset: Int64,
        connection: FramedConnection
    ) async throws {
        var offset = startingOffset
        while offset < offer.descriptor.expectedSize {
            let chunk = try await receive(from: connection)
            guard chunk.kind == .chunk,
                  chunk.requestID == requestID,
                  chunk.offset == UInt64(offset),
                  !chunk.payload.isEmpty
            else {
                throw SyncServerSessionError.invalidChunk
            }
            try await writer.append(chunk.payload, offset: offset)
            offset += Int64(chunk.payload.count)
            if offset % SyncConstants.checkpointSize == 0 {
                try await writer.checkpoint(at: offset)
            }
        }
    }

    private func failureResult(
        for error: any Error,
        retryableIntegrityFailure: Bool
    ) -> TransferResult {
        if let writerError = error as? DestinationWriterError,
           writerError == .integrityMismatch {
            return .failed(
                code: .integrity,
                message: "Received bytes failed integrity verification.",
                retryable: retryableIntegrityFailure
            )
        }
        if error is SyncServerSessionError || error is FrameCodecError {
            return .failed(
                code: .invalidFrame,
                message: "The transfer stream was invalid.",
                retryable: false
            )
        }
        return .failed(
            code: .destinationUnavailable,
            message: "The destination could not accept this resource.",
            retryable: false
        )
    }
}

private actor SessionOpeningDeadline {
    private enum State: Equatable {
        case pending
        case accepted
        case timedOut
    }

    private var state = State.pending

    var didTimeOut: Bool {
        state == .timedOut
    }

    func markAccepted() -> Bool {
        guard state == .pending else { return false }
        state = .accepted
        return true
    }

    func markTimedOut() -> Bool {
        guard state == .pending else { return false }
        state = .timedOut
        return true
    }
}

File: packages/SyncCore.Windows/src/receiver/sync-server-session.ts
// SyncServerSession — state machine that mirrors
// `packages/SyncCore/Sources/MacReceiverKit/SyncServerSession.swift`:
//
//   accept() → openSession (15s opening deadline)
//     → first frame .session(.request) + matching binding → accept
//     → loop on .offer / .session(.finished)
//         .offer → writer.begin → .adopted OR .transfer
//         .transfer → .decision(.start|.resume) → receiveBytes → commit
//         integrityMismatch (1st) → retryable; (2nd) → terminate session
//     → on .session(.finished) → return SyncSummary
//
// This is the runtime that ties ManifestStore + DestinationWriter +
// FramedConnection together. The iOS side drives the wire state machine;
// the receiver reacts.

import { EventEmitter } from 'node:events';
import {
  decodeControlMessage,
  decodeDecision,
  decodeResult,
  decodeSessionMessage,
  encodeDecision,
  encodeResult,
  encodeSessionRequest,
  FrameCodec,
  type FrameHeader,
  FrameKind,
  FramedConnection,
  SyncConstants,
  type SyncSummary,
} from '../protocol/index.js';
import { DestinationWriter, type WriterBeginResult } from './destination-writer.js';
import { ManifestStore } from './manifest-store.js';
import type { OperationLogBuffer, OperationLogLevel } from '../logging/operation-log.js';

export interface SyncServerSessionOptions {
  connection: FramedConnection;
  manifest: ManifestStore;
  writer: DestinationWriter;
  operationLog: OperationLogBuffer;
  onAccepted?: () => void;
  onEvent?: (event: { level: OperationLogLevel; category: string; message: string }) => void;
  openingTimeoutSeconds?: number;
  idleTimeoutSeconds?: number;
}

export class SyncServerSession extends EventEmitter {
  private readonly connection: FramedConnection;
  private readonly manifest: ManifestStore;
  private readonly writer: DestinationWriter;
  private readonly operationLog: OperationLogBuffer;
  private readonly onAccepted?: () => void;
  private readonly onEvent?: (event: { level: OperationLogLevel; category: string; message: string }) => void;
  private readonly openingTimeoutSeconds: number;
  private readonly idleTimeoutSeconds: number;
  private integrityFailures = new Map<string, number>();

  constructor(options: SyncServerSessionOptions) {
    super();
    this.connection = options.connection;
    this.manifest = options.manifest;
    this.writer = options.writer;
    this.operationLog = options.operationLog;
    this.onAccepted = options.onAccepted;
    this.onEvent = options.onEvent;
    this.openingTimeoutSeconds = options.openingTimeoutSeconds ?? SyncConstants.defaultOpeningTimeoutSeconds;
    this.idleTimeoutSeconds = options.idleTimeoutSeconds ?? SyncConstants.defaultIdleTimeoutSeconds;
  }

  // Receives the next frame, bounding the wait so a sender that vanishes
  // (locked iPhone, suspended app, LAN drop) ends the session instead of
  // holding the receiver's single active connection slot open forever.
  private async receive(): Promise<{ header: FrameHeader; payload: Uint8Array }> {
    let timer: NodeJS.Timeout | undefined;
    const idle = new Promise<never>((_, reject) => {
      timer = setTimeout(() => {
        this.connection.destroy();
        reject(new Error('idleTimedOut'));
      }, this.idleTimeoutSeconds * 1000);
      timer.unref?.();
    });
    try {
      return await Promise.race([this.connection.receive(), idle]);
    } finally {
      if (timer) clearTimeout(timer);
    }
  }

  async run(): Promise<SyncSummary> {
    const summary: SyncSummary = { added: 0, existing: 0, notLocal: 0, failed: 0 };
    try {
      await this.openSession();
      this.onAccepted?.();
      while (true) {
        const { header, payload } = await this.receive();
        if (header.kind === FrameKind.session) {
          const session = decodeSessionMessage(payload);
          if (session.kind === 'finished') {
            this.emitEvent('success', 'Session', `Completed: ${summary.added} added, ${summary.existing} already present, ${summary.notLocal} not local, ${summary.failed} failed.`);
            await this.sendResult({ kind: 'sessionCompleted', summary });
            return summary;
          }
          throw new Error('unexpected session message');
        }
        if (header.kind === FrameKind.offer) {
          await this.handleOffer(payload, summary);
          continue;
        }
        throw new Error(`unexpected frame kind: ${header.kind}`);
      }
    } finally {
      this.connection.destroy();
    }
  }

  private async openSession(): Promise<void> {
    const deadline = new Promise<never>((_, reject) => {
      setTimeout(() => reject(new Error('openingTimedOut')), this.openingTimeoutSeconds * 1000).unref?.();
    });
    const received = this.connection.receive();
    const { header, payload } = await Promise.race([received, deadline]);
    if (header.kind !== FrameKind.session) {
      throw new Error('protocolViolation: first frame must be .session');
    }
    const session = decodeSessionMessage(payload);
    if (session.kind !== 'request') {
      throw new Error('protocolViolation: first session message must be .request');
    }
    const accepted = this.manifest.acceptSession(session.albumID, session.albumName, session.sourceBindingID);
    this.emitEvent('info', 'Session', `Opening album "${session.albumName}".`);
    this.writer.prepareAlbumDirectory(session.albumName);
    this.emitEvent('success', 'Session', `Accepted album "${session.albumName}" in folder "${accepted.album.destinationFolderName}".`);
    await this.sendSession(encodeSessionRequest({
      albumID: session.albumID,
      albumName: session.albumName,
      sourceBindingID: accepted.sourceBindingID,
    }));
    // Note: encodeSessionRequest above is the "request" form. The Swift
    // side sends a literal {sourceBindingID} object via the
    // `.accepted(sourceBindingID)` case. We send the same shape here.
    // The actual encoding for .accepted is a one-key object; the helper
    // above is reused by relying on the .accepted branch of
    // decodeSessionMessage.
    void decodeSessionMessage;
  }

  private async handleOffer(payload: Uint8Array, summary: SyncSummary): Promise<void> {
    const offer = JSON.parse(new TextDecoder().decode(payload)) as {
      resourceID: string;
      descriptor: {
        assetLocalIdentifier: string;
        resourceType: string;
        originalFilename: string;
        duplicateOrdinal: number;
        contentHash: string;
        expectedSize: number;
        creationDate?: string;
        role?: string;
      };
    };
    const logicalResourceID = `${offer.resourceID}`; // computed identically by iOS
    this.manifest.ensureTransferRecord(offer, logicalResourceID, offer.descriptor.assetLocalIdentifier);
    this.emitEvent('info', 'Resource', `Offered "${offer.descriptor.originalFilename}" (${offer.descriptor.expectedSize} bytes).`);

    const result: WriterBeginResult = this.writer.begin({
      resourceID: offer.resourceID,
      logicalResourceID,
      albumID: offer.descriptor.assetLocalIdentifier,
      descriptor: {
        contentHash: offer.descriptor.contentHash,
        expectedSize: offer.descriptor.expectedSize,
        originalFilename: offer.descriptor.originalFilename,
        creationDate: offer.descriptor.creationDate ? new Date(offer.descriptor.creationDate) : undefined,
        role: offer.descriptor.role,
      },
    });

    if (result.kind === 'adopted') {
      summary.existing += 1;
      this.integrityFailures.delete(offer.resourceID);
      this.emitEvent('info', 'Resource', `Skipped "${offer.descriptor.originalFilename}"; already present.`);
      await this.sendDecision(FrameKind.decision, { kind: 'skip' });
      return;
    }

    if (result.offset === 0) {
      this.emitEvent('info', 'Resource', `Receiving "${offer.descriptor.originalFilename}".`);
    } else {
      this.emitEvent('info', 'Resource', `Resuming "${offer.descriptor.originalFilename}" at byte ${result.offset}.`);
    }
    await this.sendDecision(FrameKind.decision, result.offset === 0
      ? { kind: 'start', offset: 0 }
      : { kind: 'resume', offset: result.offset });

    const expectedSize = offer.descriptor.expectedSize;
    let offset = result.offset;
    const requestID = FrameCodec.newRequestID();
    while (offset < expectedSize) {
      const { header: chunkHeader, payload: chunkPayload } = await this.receive();
      if (chunkHeader.kind !== FrameKind.chunk) {
        throw new Error('protocolViolation: expected chunk');
      }
      if (chunkPayload.length === 0) {
        throw new Error('protocolViolation: empty chunk');
      }
      if (chunkHeader.offset !== BigInt(offset)) {
        throw new Error('protocolViolation: chunk offset mismatch');
      }
      this.writer.append(offset, chunkPayload);
      offset += chunkPayload.length;
      if (offset % SyncConstants.checkpointSize === 0) {
        this.writer.checkpoint(offset);
      }
    }

    try {
      const commit = this.writer.commit(offer.descriptor.contentHash);
      summary.added += 1;
      this.integrityFailures.delete(offer.resourceID);
      this.emitEvent('success', 'Resource', `Committed "${offer.descriptor.originalFilename}" to "${commit.relativePath}".`);
      await this.sendResult({ kind: 'committed', relativePath: commit.relativePath });
    } catch (err) {
      const code = (err as Error).message.includes('integrity') ? 'integrity' : 'destinationUnavailable';
      const failures = (this.integrityFailures.get(offer.resourceID) ?? 0) + 1;
      this.integrityFailures.set(offer.resourceID, failures);
      const retryable = code === 'integrity' && failures === 1;
      this.emitEvent(
        retryable ? 'warning' : 'error',
        'Resource',
        `Failed "${offer.descriptor.originalFilename}": ${(err as Error).message}${retryable ? ' Retrying once.' : ''}`,
      );
      await this.sendResult({ kind: 'failed', code: code as 'integrity' | 'destinationUnavailable', message: (err as Error).message, retryable });
      if (!retryable) {
        summary.failed += 1;
        throw new Error('integrityFailureLimitExceeded');
      }
    }
  }

  private async sendDecision(kind: FrameKind, decision: { kind: 'skip' } | { kind: 'start' | 'resume'; offset: number }): Promise<void> {
    void decodeDecision;
    const payload = new TextEncoder().encode(encodeDecision(decision));
    await this.connection.send({
      kind,
      requestID: FrameCodec.newRequestID(),
      offset: 0n,
      payload,
    });
  }

  private async sendResult(r: { kind: 'committed'; relativePath: string } | { kind: 'failed'; code: 'integrity' | 'destinationUnavailable'; message: string; retryable: boolean } | { kind: 'sessionCompleted'; summary: SyncSummary }): Promise<void> {
    const payload = new TextEncoder().encode(encodeResult(r));
    await this.connection.send({
      kind: FrameKind.result,
      requestID: FrameCodec.newRequestID(),
      offset: 0n,
      payload,
    });
  }

  private async sendSession(payload: string): Promise<void> {
    await this.connection.send({
      kind: FrameKind.session,
      requestID: FrameCodec.newRequestID(),
      offset: 0n,
      payload: new TextEncoder().encode(payload),
    });
  }

  private emitEvent(level: OperationLogLevel, category: string, message: string): void {
    this.onEvent?.({ level, category, message });
  }
}

export type { SyncSummary };
